### PR TITLE
Improved makeOverridable

### PR DESCRIPTION
**Pull request by ElvishJerricco**
Motivation for this change: As explained in #26561, using `haskellPackages.extend` breaks `haskellPackages.override`, meaning there is no way to override the arguments passed to `haskellPackages`, such as `all-cabal-hashes`. This PR introduces a much more powerful notion of overriding into `makeOverridable` so that the solution to this problem can be enjoyed by everything.

### CODE REVIEW

File: lib/fixed-points.nix [hunk 1]
```diff
@@ -71,8 +71,19 @@ rec {
...
+  # Similar to `makeExtensible`, but expects you to implement the
+  # final interface for the result. Specifically, it takes an extra
+  # argument: a function that takes the final result and the `extend`
+  # function as arguments, and returns a transformed result
+  # (preferably one that contains the `extend` function). This is
+  # mainly useful for getting to choose what to name the `extend`
+  # function in the resulting attribute set. But it's also useful for
+  # having an internal structure that extensions can see, but the user
+  # facing code cannot.
+  makeExtensibleWithInterface = interface: fext: interface
```
Comment (Ericson2314) [inline lib/fixed-points.nix]: @cstrahan I need to write something longer up another time, but I want to close my PR in lieu of this generalization.

File: lib/customisation.nix [hunk 2]
```diff
@@ -51,9 +51,46 @@ rec {
        else { }));
 
 
-  /* `makeOverridable` takes a function from attribute set to attribute set and
-     injects `override` attibute which can be used to override arguments of
-     the function.
+  # Like `makeOverridable`, but provides the function with the `self`
+  # argument. `f` is called with the new `self` whenever an override
+  # or extension is added.
+  makeOverridableWithSelf = f: origArgs: let
```
Comment (Ericson2314) [inline lib/customisation.nix]: I think `super` is actually the new argument, so this needs to be renamed.

**Comment by Ericson2314** [comment]
@ElvishJerricco Looking good! Some of the complexity of the `*scope` stuff I think @nbp and @peti want to remove altogether, but that can be done later of course.

I think `iself` and `WithSelf` should become `output` and `WithOutput`---I don't want to confuse people over `iself` vs `self`.

**Comment by ElvishJerricco** [comment]
@Ericson2314 yea that sounds like a mostly better name, although `output` is generally a term that refers to Nix store outputs.

File: pkgs/development/haskell-modules/default.nix [hunk 3]
```diff
@@ -7,26 +7,24 @@
 , configurationNix ? import ./configuration-nix.nix
 }:
 
+self: # Provided by `callPackageWithSelf`
```
Comment (Ericson2314) [inline pkgs/development/haskell-modules/default.nix]: This would be the final output, right?

File: pkgs/development/haskell-modules/default.nix [hunk 4]
```diff
@@ -7,26 +7,24 @@
...
-            (extends nixConfiguration haskellPackages)))));
-
-in
-
-  extensible-self
+in (extends overrides
+     (extends packageSetConfig
+       (extends compilerConfig
+         (extends commonConfiguration
+           (extends nixConfiguration haskellPackages))))) self
```
Comment (Ericson2314) [inline pkgs/development/haskell-modules/default.nix]: That seems like the wrong thing to put there though?

Comment (Ericson2314) [inline pkgs/development/haskell-modules/default.nix]: I'll need to think about it.

File: pkgs/development/haskell-modules/default.nix (hunk 3 above)
Comment (ElvishJerricco) [inline pkgs/development/haskell-modules/default.nix]: Yea. Since `self` becomes the scope for all the packages' `overrideScope`, it needs to support `extend`, thus it needs to be the final output, not the internal one.

File: pkgs/development/haskell-modules/default.nix (hunk 4 above)
Comment (ElvishJerricco) [inline pkgs/development/haskell-modules/default.nix]: Are you talking about the `self`? There's no reason it *shouldn't* be the final output, and I didn't want to have to pass around two different `self`s

File: lib/customisation.nix [hunk 5]
```diff
@@ -109,12 +196,34 @@ rec {
...
+    let f = if builtins.isFunction fn then fn else import fn;
+    in callPackageCommon (builtins.functionArgs f) autoArgs (iself: x: _: f x) args;
+
+
+  # Like `callPackageWith`, but provides the function with the `self`
+  # argument. `fn` is called with the new `self` whenever an override
+  # or extension is added.
+  callPackageWithSelfWith = autoArgs: fn: args:
+    let f = if builtins.isFunction fn then fn else import fn;
+    in callPackageCommon (builtins.functionArgs f) autoArgs (iself: args: _: f args iself ) args;
```
Comment (Ericson2314) [inline lib/customisation.nix]: Right  I think the last parameter should also be passed in----it is what mk-haskell-packages is supposed to feed into the `(extend _ (extend _ ...))` chain.

Comment (ElvishJerricco) [inline lib/customisation.nix]: I don't know if that's true. It definitely needs the final output for the scope, which is sort of expected to be the same thing as the `self` in the extensions chain.

**Comment by cstrahan** [comment]
Oh, I was unaware of this work :). I'll look when I'm back at my desk in a couple hours.

**Comment by ElvishJerricco** [comment]
Wait... Is that travis failure a memory leak in the Nix expressions?... Did I do that?

**Comment by cstrahan** [comment]
@ElvishJerricco  So, I need a little bit more time to think this over. In the interim, I'd like to point out -- and get your thoughts on -- another solution I played with: https://github.com/NixOS/nixpkgs/pull/15095#issuecomment-216114459

There might be something there that you can draw inspiration from.

**Comment by cstrahan** [comment]
Alright, I've looked through this pretty extensively, and it all looks good to me.

I see from your example:

```
obj.overridePackage (output: self: super: { args = super.args // {b = self.val.a;}; })
```

It's apparent that `overridePackage` can serve the purpose of both `override` and `extend`, and solves the issue I was trying to address in https://github.com/NixOS/nixpkgs/pull/15095#issuecomment-216114459 where I wanted a way to allow for extensions to refer to the package/package-set's original input arguments, at the expense of being a bit more verbose than either. And for completeness, it looks like you could also elide attrs in either case, should the need arise. I suspect that `override` and `extend` will suit 99% of use cases, while `overridePackage` makes possible that last 1%, so I think the flexibility justifies the interface. Very nice work!

I don't want to bike-shed too much, but I do wonder if we can come up with more intuitive name. I've been pushing for a change in vernacular for a while now, where `override` would indicate changing input arguments whereas `extend` allows for modification of a resulting package/-set. I don't know where that leaves us for naming what is essentially both in one, though. Considering that it applies equally well to both individual packages and sets of packages, the "Package" in `overridePackage` might be a little misleading.

**Comment by cstrahan** [comment]
Another thing I'm excited about is being able to use `extend` with `mkDerivation`. `overideDerivation` doesn't compose, and `overrideAttrs` doesn't give you the fixedpoint, so having `override`+`extend`+`overridePackage` is a huge step up.

**Comment by ElvishJerricco** [comment]
@cstrahan Thanks =) I agree about the naming of `overridePackage`; I think it's definitely the wrong name. I'm just not sure how to encapsulate the idea that it can both override and extend in one name. I guess the easy answer would be `overrideAndExtend` =P

**Comment by nbp** [comment]
@ElvishJerricco 

I like the fact that we have both the set of arguments (`args`) and the value (`val`) which overridable in one function, as this follow the idea that @peti and I discussed at the FOSDEM for pushing the mutation of these under the same operator ( https://github.com/NixOS/rfcs/pull/3 )

Still, I fail to understand the need for `output`, and why it would be different than `self`, can you clarify a bit more?  `self` is supposed to be the latest state, so `output` sounds redundant, or this is a bug in the implementation of `self`.

One of the comment describes `self` being the state of one package before any override.  Than is not supposed to be its meaning.  Is this a desired feature, to have access to the package before any of its override?  Why should we privileged the first implementation over the overridden implementation of a package?

**Comment by Ericson2314** [comment]
I wonder if rather than supporting `extend` directory, `makeOverridable` should just detect and wrap it like it does with `overrideAttr`s.

First of all, note that `overrideDerivation` should really put on by `mkDerivation`, not `makeOverridable`/`callPackages`. Then we have:

- `override`: Override args to `callPackage`d function.
- `overrideAttrs`: Override args to `mkDerivation`.
- `overrideDerivation`: Override final value.

In general, one can imagine a bunch of functions composed such that the output of one is the input of the next. Each argument set and the final value may also be a fixed point. These combinations stem from a grammar like
```
overridable-thing ::= any-type-not-fixed
                   |  any-type-fixed-point 
                   |  attrset -> overridable-thing
```

@ElvishJerricco's work is strictly better than what exists so far, but handles just the `attrset -> (attset-fixed-point | any-type-not-fixed` cases. I'd ideally like to follow the grammar and make `makeOverridable` responsible for the entire `attrset -> overridable-thing` non-terminal.

**Comment by cstrahan** [comment]
@nbp 

> Still, I fail to understand the need for output, and why it would be different than self, can you clarify a bit more? self is supposed to be the latest state, so output sounds redundant, or this is a bug in the implementation of self.

As I understand it (and perhaps @ElvishJerricco can correct me if I've got any of this wrong), the idea is that the `output` is a projection of the usual fixpoint. Now, when you `override/`extend`/`overridePackage` the set, you're tweaking the underlying fixpoint, and you get a new resulting projection. This allows for a different interface to be exposed to the end user rather than the inner fixpoint.

One real-world-ish example where this could be useful is when your package set includes a package with a name that conflicts with the typically added `override`/`extend` attributes, so you might want to expose those functions under different names.

More importantly, this separation is leveraged in the implementation: `makeOverridable` now has as its inner fixpoint `{ args = <...>; val = <...>; }`, which is what you have access to when you call `overridePackage`:

```
obj.overridePackage (output: self: super: { args = super.args // {b = self.val.a;}; })
```

From the outside, a user of a package[-set] constructed via `makeOverridable` will see the nice "interface", that is, the typical derivation consisting of `{ name = ...; src = ...; ...etc... }`.

So far, I've only justified the separation of that inner fixpoint from the exposed interface, but I haven't justified passing around `output` yet (which, keep in mind, is that `{ name = ...; src = ...; ...etc... }` attrset mentioned in the last paragraph). Passing `output` is important when it comes to uses of `callPackageWithOutput`, as seen in the Haskell package set changes in this PR. The packages within `haskellPackages` need to refer to each other, so that means they either need to refer to the "behind-the-scenes" fixpoint (`{ args = <...>; val = <...>; }`) or they can use the intuitive projection thereof (that is, what the end user will see). I think it makes sense to thread through the latter; from the point of view of `haskellPackages` the inner `{ args = <...>; val = <...>; }` fixpoint is just an implementation detail of `makeOverridable`.

You can see where `output` is threaded through for the above purposes here:

```
  callPackageWithOutputWith = autoArgs: fn: args:
    let f = if builtins.isFunction fn then fn else import fn;
    in callPackageCommon (builtins.functionArgs f) autoArgs (output: args: _: f args output ) args;
```

`f args output` is where e.g. the resulting interface is pushed into the `haskellPackages` set.

**Comment by ElvishJerricco** [comment]
@Ericson2314 It's unclear to me what exactly that grammar means and what exactly you're proposing. My intuition around this `makeOverridable` stuff is centered around functions: `args -> val`. The goal is to put both the arguments and the output value into a context where they can both be observed and changed arbitrarily together. The important semantic is that the fixed point begins with `val = f self.args`. Your grammar only seems to talk about `args`?

**Comment by nbp** [comment]
@cstrahan
> As I understand it (and perhaps @ElvishJerricco can correct me if I've got any of this wrong), the idea is that the output is a projection of the usual fixpoint. Now, when you override/extend/overridePackage the set, you're tweaking the underlying fixpoint, and you get a new resulting projection. This allows for a different interface to be exposed to the end user rather than the inner fixpoint.

There is no such thing as outer / inner fix-point, except added in these patches.  The reason why it is added is totally obscure to me.  Making relations between the final set of packages should be done through `self`, which is supposed to take the latest state of the packages.  Adding an additional fix-point on top of `self` does not add any new feature which is not supported by the `self` fix-point already.

I do not understand why we need this extra fix-point.  If this work would be rewritten without this extra fix-point then I would more inclined to accept it.

**Comment by ElvishJerricco** [comment]
@nbp it would not be possible to both provide an interface without the `{ args, val }` structure, and provide access to the `{ args, val }` structure for overriding without separating the concepts of the in Dr and outer fixed points. Having that structure for overriding is essential to unifying the concepts of `override` and `extend`

**Comment by nbp** [comment]
@ElvishJerricco even if we replace the `self` / `makeExtensible` from `haskell-modules/default.nix` and use one computed from the fix-point of Nixpkgs ? Maybe making `top-level/haskell-packages.nix` a proper overlay would help.

**Comment by ElvishJerricco** [comment]
@nbp that's not really a general enough solution. Not all things live in a fixed point (not all things live in nixpkgs). Plus, if we relied on layering fixed points, we'll end up back with the original problem: using `override` will wipe any changes made via overlays, and generate inconsistency as the `self` passed in will still be the old self from before these new overrides but with those overlays applied. You can't layer the fixed points without losing consistency somewhere.

**Comment by nbp** [comment]
@ElvishJerricco If I understand correctly, you are saying that you want to add another fix-point because people doing override out-side of Nixpkgs cannot get their overriden packages used as input of other haskell packages because they are taken out of self, and not whatever overriden set they end-up with?

So I have one question, is it possible or not to make an override in an overlay, and make it used as input of other haskell packages?  If not, this is what we should be fixing then.

**Comment by ElvishJerricco** [comment]
> If I understand correctly, you are saying that you want to add another fix-point because people doing override out-side of Nixpkgs cannot get their overriden packages used as input of other haskell packages because they are taken out of self, and not whatever overriden set they end-up with?

No. When you use `haskellPackages.extend`, it returns the result from *before* `makeOverridable` gets to add its overrides back in. Thus you cannot chain `override` and `extend` arbitrarily, as `extend` will cease your ability to `override`. This incompatibility is inherent in the layered fixed point design. It can only be solved by unifying the concept of `extend` and `override`.

> So I have one question, is it possible or not to make an override in an overlay, and make it used as input of other haskell packages? If not, this is what we should be fixing then.

This is exactly what `extend` currently does. But as I said, this eliminates the ability to use `override`. This PR is the solution to that problem.

**Comment by nbp** [comment]
@ElvishJerricco There is no issue for using `override` in Nixpkgs, only in the Haskell packages because it is using these `extend` functions within Nixpkgs.

Would it be possible to replace the `extend` from Haskell by the `extend` used by Nixpkgs stages (overlays), thus using the `override` within the Nixpkgs fixpoint, instead of out-side Haskell packages fix-point? Basically making Haskell packages become an overlay within the Nixpkgs fix-point.

**Comment by ElvishJerricco** [comment]
I don't think that's a god idea. Using `extend` outside of an overlay is a very common use case. Also, `haskellPackages` is not defined at the point where most people use it. They use `nixpkgs.haskellPackages`, but it's defined at `nixpkgs.haskell.packages.ghcXXX`. For people to extend as you suggest, they must know exactly which package set their using and what to modify to propagate the changes correctly. That's non-composable, since a common use of overlays is to change the package set at `nixpkgs.haskellPackages`. Plus, it's still inconsistent. If you use `haskellPackages.override` from outside an overlay, all of the packages in the package set will still be based on the old package set, which is wrong.

**Comment by cstrahan** [comment]
> There is no issue for using override in Nixpkgs, only in the Haskell packages because it is using these extend functions within Nixpkgs.

It should be noted that it's *more* than just Haskell packages that support extend: this also applies to Erlang and Kernel packages. For all three of the above, `override` and `extend` don't compose at the moment. It's really unintuitive, tricky behavior and we ought to fix it.

I agree with @ElvishJerricco's conclusion that `override` and `extend` need to be unified in such a way that one doesn't clobber the other.

> There is no such thing as outer / inner fix-point, except added in these patches.

That's true. However, I'm not convinced that adherence to existing architecture is a persuasive argument against a change set.

Here's my thinking:

1. The API exposed (the `override`, `extend` and `overridePackage` functions)  isn't controversial; `override` and `extend` should compose. Additionally, `overridePackage` provides a means of extending while *also* referring to the original inputs, which is presently sorely lacking.
2. It's quite possible that some things could be cleaned-up and/or made simpler, but I haven't seen any concrete suggestions. Assuming we're in agreement on the preceding point, I think we should merge what @ElvishJerricco has developed thus far, and an API compatible refactoring can take place if/when someone has the appropriate insight to do so.

If you disagree with either points, @nbp, I'd appreciate your feedback.

I'm all for merging this, and would do so myself, but I would like to first see more approval. In part because I don't want anyone to feel like I'm steamrolling them, and also in part because I'm open to being persuaded away from my present stance on this PR.

**Comment by Ericson2314** [comment]
I added this to the `17.09` milestone---the situation should somehow be improved by then.

**Comment by nbp** [comment]
> I agree with @ElvishJerricco's conclusion that `override` and `extend` need to be unified in such a way that one doesn't clobber the other.

> `override` and `extend` should compose.

No, there should not be any need for that. If there is such a need, this is a problem related to the way these packages sets are made, and I highly prefer to converge toward a uniform way of writing packages than inventing new way to get around every time.

My take is that this issue is solving the wrong problem.  `extend` are working at the top-level and this should never interfere with `override`.  If you use `extend` within an attribute of a package set (`hackellPackages` for example) then you are doing it wrong.

The `extend` function should be used around a package set, and not within a package set, like the staging mechanism of Nixpkgs do.

The  `override` functions are within the package set, and using the fix-point provided around the package set, and there is no conflict this way.

If anybody needs guidance to fix `haskellPackages` set I can help with that, but from my point of view these changes are not the right answer, and we should not accept these.

Basically, what I am asking is that the haskell packages to be converted into an overlay which does not make use of `extend` within a package set. Around, yes, but not inside!

**Comment by ElvishJerricco** [comment]
@nbp I could be misunderstanding your comment, but the problem this PR solves isn't that you couldn't use `extend` inside the package set. The problem that currently exists is that anyone with a `haskellPackages` is not able to call `extend` on it without then losing the `override` function. The `extend` that currently exists on `haskellPackages` is just broken. The changes made in this PR are necessary to fix it, because the current model for `extend` in `haskellPackages` ignores `override` entirely.

That said, I will agree that it ought to be inadvisable for a package set's definition to make use of its own `extend`, so I could see an argument for removing the "output" stuff in `callPackageWithOutputWith` from this PR.

**Comment by cstrahan** [comment]
@nbp 

>> override and extend should compose

> My take is that this issue is solving the wrong problem. extend are working at the top-level and this should never interfere with override. If you use extend within an attribute of a package set (hackellPackages for example) then you are doing it wrong.

That's _not_ what is meant by "override and extend should compose", and not what this PR is about. Whether one should be able to reference `override` and/or `extend` of a given package set _within_ that same package set is debatable, but again, that's _not_ what this PR is primarily addressing.

> Basically, what I am asking is that the haskell packages to be converted into an overlay which does not make use of extend within a package set. Around, yes, but not inside!

We're not interested in using `extend` of the surrounding package set _within_ the same package set, rather something else entirely.

Consider the following:

(NOTE: I chose the following one-liner for the sake of brevity and concision; a more likely variation would be where you're using an overlay that provides the former half (e.g. because you want a set of known-good Haskell packages suitable for doing GHCJS development with [reflex-frp](https://github.com/reflex-frp/reflex)), and that overlay is provided by someone else in some separate repo, and now in your own project's nix expression, you want to `.override` the overlayed package set such that it uses a different stdenv throughout. To provide a full fledged, real world example like this would require I set up a couple repos and many files and it would generally distract from the real meat of the matter.)

`(haskellPackages.extend my-extension).overide { stdenv = my-stdenv; }`

What should the result be? I would expect a package set with the extensions from `my-extension` _and_ the [original `stdenv` argument](https://github.com/NixOS/nixpkgs/blob/4565378641e668a15ba0d03a6ecec297551c74f9/pkgs/development/haskell-modules/default.nix#L17) to be overridden with `my-stdenv`. Right now in nixpkgs, the best you could get is the _original_, non-extended/non-overlayed `haskellPackages` with `stdenv` set to `my-stdenv`. This applies to all package sets in nixpkgs that provide `extend` and `override` (of the top of my head: erlang packages, kernel packages, haskell packages).

If you flip things around:

`(haskellPackages.overide { stdenv = my-stdenv; }).extend my-extension`

Now you'll have a package set with the extensions, but _without_ the overridden `stdenv`. Again, at the present moment, _all_ package-sets in nixpkgs have this problem.

That's what's meant by "override and extend should compose". If you'll forgive my slight abuse of mathematical notation, this is what we have _right now_ in nixpkgs (where o=override, e=extend, p=package-set):

    e ∘ o ∘ p ≡ e(p(x))
    o ∘ e ∘ p ≡ o(p(x))

If you get a package set from someone else, you'd think it'd be safe to apply subsequent `.override` or `.extend`, but that's _not_ a safe assumption right now, as the package set provided by your overlay (or package function, or whatever sort of intermediate abstraction) will "forget" everything. There's no need for composition here to deviate from the standard mathematical notion:

    f ∘ g ≡ f(g(x))
    ; and thus, for free
    f ∘ g ∘ h ≡ f(g(h(x)))

This PR resolves this problem.

Fundamentally, any solution to this problem will logically _require_ an implementation of `override` and `extend` that are mutually aware of each other. I suspect that's obvious, but I can expound here if needed.

**Comment by ElvishJerricco** [comment]
>     (haskellPackages.extend my-extension).overide { stdenv = my-stdenv; }

> [...] Right now in nixpkgs, the best you could get is the original, non-extended/non-overlayed haskellPackages with stdenv set to my-stdenv. [...]

> If you flip things around:

>     (haskellPackages.overide { stdenv = my-stdenv; }).extend my-extension

> Now you'll have a package set with the extensions, but without the overridden stdenv. [...]

While you are correct that this is where the problem lies, I should point out that the symptoms of breakage are a little different than you've described. Namely, the former line will simply fail, and the latter will work "correctly." The symptom is that whenever you use `extend`, it returns a package set that no longer has the `override` function, or any of the other work that might have been done by `makeOverridable`. Instead, it returns the internal view of the package set. So you can use `override` and *then* `extend`, but then you can no longer use `override`, because you can never do `extend` then `override`.

**Comment by cstrahan** [comment]
> While you are correct that this is where the problem lies, I should point out that the symptoms of breakage are a little different than you've described.
> [...]
> Namely, the former line will simply fail [...]

Right, exactly. That's why I said "the best you could get [...]" in my comment, but I probably should have explicitly pointed out that this use case will currently throw before it could even get that far.

> [...] and the latter will work "correctly."

Hm... Right, I think all uses of `makeExtensible` might allow for `override` followed by `extend` in a way that works correctly (without scare quotes), though not the reverse. Too busy at the moment to test that.

**Comment by ElvishJerricco** [comment]
I've fixed the merge conflicts with this branch. What can I do to try to get this or something like it into 17.09?

**Comment by nbp** [comment]
@cstrahan Thanks for your examples, but that does not change my mind on refusing these changes. 


I discussed a bit with @peti, which roughly explained me that `.extend` is mostly used as a way to duplicate a package set with a different version, and replicate these modifications to the set of inputs of scoped packages.  Basically, this is a recursive `.override` on packages which carry the scope information of where it is coming from.  If we had a function for "re-scoping the dependencies used as inputs of a set of packages", and a function to "apply override recursively to a set of packages", then we could re-implement the `extend` function, except that we would not have the extra re-scoping that you are complaining about.

This is why I consider `extend` to be the problem, and thus this solution to be a work-around which adds more complexity, than what a proper fix, splitting the `extend` function would do.

**Comment by shlevy** [comment]
@ElvishJerricco Haven't had a chance to look through the code in detail yet, so maybe this already exists, but can you provide a real world example of something you need to do but can't without this change?

@nbp If you're opposed to this change, can you suggest an alternative solution to the real world example @ElvishJerricco gives?

I'll review this PR in detail later this week, but answers to these will make it easier.

**Comment by ElvishJerricco** [comment]
> Basically, this is a recursive .override on packages which carry the scope information of where it is coming from.

I disagree. This is about functions `args -> val`, and how we manipulate bot the inputs and the outputs of such functions. At a certain level, it has nothing to do with package sets like `haskellPackages`. Without these changes, `makeOverridable` makes the arbitrary choice that we should only be able to manipulate the arguments to the function (by way of `override`), and that outputs should simply not be manipulated. This has served us well for simple derivations, but fails for anything whose output is more meaningful than a derivation. This is evidenced by the fact that `makeOverridable` adds `overrideDerivation`, which is actually an output manipulation that has been blessed with a magic implementation.

This change is about being able to manipulate both the arguments *and* the outputs of these functions; e.g. `overrideDerivation` actually becomes something that anyone could have defined, instead of the magic function it currently is. This happens to apply nicely to package sets like `haskellPackages` because it means we can use `override` to change the package set level inputs like GHC, and `extend` to change the contents of the package set. But the purpose is more general and fairly simple. If I'm understanding correctly, what @nbp is describing is to leave `makeOverridable` the way it is, and to instead have a different mechanism entirely specifically for package sets. Satisfying the full potential of the existing design of `makeOverridable` by removing its restriction to arguments seems much better to me.

---

As for some concrete examples: `haskellPackages` has an `overrides` input, which you can change using `override` to get something similar to `extend`. `reflex-platform` uses this field to define its haskell overrides. The problem is that the package set defined by `reflex-platform` is meant to be a drop in replacement for `haskellPackages`, and `reflex-platform` wants to compose several overrides. So, the fact that someone can't just do `_.override { overrides = ...; }` without undoing and breaking the Reflex overrides fails the goal of being a drop in replacement. The solution is [this unfortunate thing `makeRecursivelyOverridable`](https://github.com/reflex-frp/reflex-platform/blob/develop/default.nix#L189), which special cases on the `overrides` argument and composes them manually.

I discovered this web of problems when I tried to refactor `reflex-platform` to use the more correct `extend`, and realized that the same problem applies to *any* nix expression you might define with `callPackage`/`makeOverridable`. If I were to refactor to `extend`, it would be a breaking change for any user of `reflex-platform` that then uses `override` on the package set. Another example could be `llvmPackages`. If I wanted to override the `llvm` version used by the other packages defined in that struct, it'd be great if I could do so using `extend`. But as is, there's no way to do this and keep everything consistent with each other short of manually overriding each package and building a new struct out of them. I can imagine a variety of struct-like things, which aren't really package sets but do have more interesting structure than a simple derivation, where compatible `override` and `extend` functions would be useful.

**Comment by shlevy** [comment]
IMO we should go ahead with this. @nbp Are you able to give at least a sketch of how you would solve @ElvishJerricco's reflex example or even @cstrahan's toy examples? If not I think we should merge for now and revisit later if something better becomes available.

**Comment by shlevy** [comment]
@ElvishJerricco If @nbp isn't able to give anything in a few days ping me after fixing conflicts and I'll merge.

**Comment by Ericson2314** [comment]
@shlevy n.b. This would be quite useful for at least cleaning up reflex-platform and making it more easily consumable down, so we might consider cherry-picking to 17.09 too.

**Comment by cstrahan** [comment]
@nbp

> @cstrahan Thanks for your examples, but that does not change my mind on refusing these changes.

I don't see where you answered my questions; they weren't rhetorical.

I feel like we're putting in an disproportionate amount of effort in communication here: we've provided illustrative example code, real world uses cases, and have tried to pin down a common vernacular to disambiguate some of the subtleties of both the problem space in general and this particular solution. At least to me, it would appear that your responses could be summarized as "I don't like how X solves problem Y", where neither X nor Y seem to describe the problem nor solution, respectively.

For your criticism of this PR to be constructive, you'll need to

* Convince us that you understand what we perceive to be the problem.
* If you disagree with what we perceive as the underlying problem, that's fine, but as per the previous bullet, you'll need to first convince us that you understood our position; then a description of what you perceive to be the true problem may follow.
* Provide sample code demonstrating at least what the surface API should look like, and what laws/invariants should hold. In addition, an example usage of your proposed alternative API should be given that clearly demonstrates that it meets the needs of the use cases we've put forward. I would suggest the `haskellPackages` use in `reflex-frp` *and* the use of `overridePackage` described in https://github.com/NixOS/nixpkgs/pull/27319#issuecomment-315500024 at minimum.

It's quite possible that this PR could be improved, and it would be awesome if your feedback could help us get there. However, at the moment, it feels like all we're managing to do is go around in circles.

**Comment by nbp** [comment]
As I mentioned before, if this does not belong in Nixpkgs, then this can go in an overlay [1], where you are free to do anything, and where this will not add burden to maintain Nixpkgs.

[1] https://github.com/mozilla/nixpkgs-mozilla/blob/master/lib-overlay.nix

**Comment by edolstra** [comment]
I'm very reluctant to merge this because a) it makes a pretty non-trivial change to a core function; b) changes to `makeOverridable` may have a significant performance / memory impact; c) I have no idea what it does or why; d) I don't have time to review it at the moment. Sorry about that.

**Review by peti — COMMENTED** [review]
FWIW, I'll provide my (somewhat unhelpful) feedback, too. I haven't had the time to read, understand, and think about these modification in detail. I find it difficult to predict what kind of consequences these changes will have. In the past, we've merged changes to this code to fix problem A (and did), but then problem B emerged, and this chain of events has been going on like that for the last 2 years or so. It seems to me like someone will have to take a step back and think about overriding in Nix one more time, considering the big picture rather than trying to solve just one particular issue we recognize at this very moment. I might actually give it a shot, but it's not an easy problem as you all know.

Anyhow, I won't merge this PR because I don't have the problem that this code wants to solve. I realize some kind of  problem exists, but I just don't have it in practice, so I'm not motivated to get into the details of this PR.

At the same time, if someone else desperately wants this merged, then I won't object either.

**Comment by shlevy** [comment]
Can someone point out an actual problem with this code? Something that will break, something that will slow down, something that will make future code harder to reason about, anything?

We have several people who have taken the time to understand this code who think it's a good solution to real-world problems, a few people who have reviewed it who aren't sure it's the best solution but haven't proposed another, and a few people who haven't reviewed it in detail but don't like it anyway. Why shouldn't we merge in this case? `git revert` is very cheap.

**Comment by Ericson2314** [comment]
I should also point out that the code is structured so that each commit stands alone is readable. @ElvishJerricco has specifically done this to make this otherwise complex change decipherable.

**Comment by shlevy** [comment]
BTW, this is oversimplified but the ultimate theoretical justification for this change is the same as the justification for `protected` class members: We have details we want to encapsulate from consumers of our interface but make available to extenders of it.

**Comment by ElvishJerricco** [comment]
@shlevy Merge conflicts are fixed btw.

**Comment by shlevy** [comment]
@edolstra Please let me know if your objection is a hard "no" to this being merged, otherwise I'll merge soon.

**Comment by cstrahan** [comment]
To touch on the performance concerns: if we don't have a readily available means of profiling built into Nix, that ought to be a priority for us, rather than punting on design improvements.

**Comment by LnL7** [comment]
This broke evaluation of `pkgs/top-level/release.nix`

**Comment by shlevy** [comment]
@LnL7 https://github.com/NixOS/nixpkgs/commit/751d397ad486a84b93907345124f550e3b32427c , thanks

File: lib/customisation.nix [hunk 6]
```diff
@@ -50,10 +50,50 @@ rec {
...
+      }) // {
+        inherit overridePackage;
+
+        override = newArgs: overridePackage (_: self: super: {
+          args = super.args //
+            (if builtins.isFunction newArgs then newArgs super.args else newArgs);
+        });
+      };
+
+  in lib.makeExtensibleWithInterface (x: o: interface (addOverrideFuncs x o) o) (output: self: {
```
Comment (LnL7) [inline lib/customisation.nix]: `attribute 'makeExtensibleWithInterface' missing, at ~/nixpkgs/lib/customisation.nix:84:6`

**Comment by LnL7** [comment]
@shlevy looks good now

**Comment by shlevy** [comment]
Tarball is broken still, fixing...

**Comment by shlevy** [comment]
Done

**Comment by peti** [comment]
This PR breaks the evaluation of `callPackage ./foo {}` where `foo/default.nix` requires a `stdenv` argument. For example, I have  in my `~/.config/nixpkgs/config.nix`:

    haskellPackages = super.haskellPackages.override {
      overrides = self: super: {
        Cabal-git = self.callPackage /home/simons/src/cabal/Cabal {};
        cabal-install-git = (self.callPackage /home/simons/src/cabal/cabal-install {}).overrideScope (self: super: {
          Cabal = self.Cabal-git;
          hackage-security = pkgs.haskell.lib.dontCheck super.hackage-security;
        });
      };
    };

and `/home/simons/src/cabal` is a recent checkout of git://github.com/haskell/cabal.git. This used to work fine, but now I get the following error:

~~~
error: anonymous function at /home/simons/src/cabal/cabal-install/default.nix:1:1 called without required argument ‘stdenv’, at /home/simons/.nix-defexpr/lib/customisation.nix:218:76
~~~

git bisect says that e11dbc37103687fb0c57dce85209086b0d691a34 is the first commit to cause that failure.

**Comment by shlevy** [comment]
@ElvishJerricco @Ericson2314  Any idea here?

**Comment by edolstra** [comment]
Please revert. This makes an already bad idea (`makeOverridable`) even more complicated. For example, the full description of `makeOverridableWithInterface` is:
```
  # A more powerful version of `makeOverridable` with features similar
  # to `makeExtensibleWithInterface`.
```
WTF does that mean? What is `makeExtensibleWithInterface`? It's not like the code is self-explanatory, with all the selfs and supers and functors and dynamic attributes etc. 

Regarding performance metrics, we have those, and they show a significant spike starting today: https://hydra.nixos.org/job/nixpkgs/trunk/metrics#tabs-charts. E.g. `nix-env.qa.allocations` went from 374 MB to 631 MB.

The claim that reverts are cheap is false in my experience; once changes like these are in and other commits start building on them, it quickly becomes very hard to untangle them.

**Comment by shlevy** [comment]
Reverting

**Comment by shlevy** [comment]
Done

**Comment by ElvishJerricco** [comment]
Fair. I understand that this change was complicated. But this still leaves us with issues like #26561. `extend` is supposed to be a fundamental way to mess with `haskellPackages`, but it's just broken. What is the correct solution for that? How else do we solve all the various problems outlined in this discussion?

**Comment by peti** [comment]
The existence of an `overrideScope` attribute that creates a complete, modified copy of the package set for the context of the given attribute may have been a mistake in the first place. One might very well argue that this method should exists only for the package set, but not for every package inside of it. It's my understanding that this interpretation is also in-line with the way overlays work.

Anyhow, technical discussion of this issue probably belongs into an RFC, not into a pull request -- unless someone feels like spending a lot of effort creating a PR that (a) works, (b) does not affect performance nor memory requirements, (c) is well documented, and (d) works out of the box without breaking evaluation of existing code.
